app/models/bulk_upload.rb

```diff
@@ -0,0 +1,193 @@
+class BulkUpload
+  include ActiveModel::Model
+  include ActiveModel::Validations
+  include ActiveModel::Conversion
+
+  SPREADSHEET_CONTENT_TYPES = %w[
+    application/vnd.ms-excel
+    application/vnd.openxmlformats-officedocument.spreadsheetml.sheet
+  ].freeze
+
+  FIRST_DATA_ROW = 7
+
+  def initialize(file, content_type)
+    @file = file
+    @content_type = content_type
+  end
+
+  def process
+    return unless valid_content_type?
+
+    xlsx = Roo::Spreadsheet.open(@file, extension: :xlsx)
+    sheet = xlsx.sheet(0)
+    last_row = sheet.last_row
+    if last_row < FIRST_DATA_ROW
+      errors.add(:case_log_bulk_upload, "No data found")
+    else
+      data_range = FIRST_DATA_ROW..last_row
+      data_range.map do |row_num|
+        case_log = CaseLog.create
+        map_row(sheet.row(row_num)).each do |attr_key, attr_val|
+          begin
```

Comment:
This is horrendously slow but the idea is to save all valid data, and not error when you hit an invalid bit. Not sure it can be easily parallelized (like this) because all the things here will be acting on the same table row... Anyone got any better ideas?